### DISCUSSION TITLE
Trouble rotating an Ellipse by Extending Scatterplot Layer

### DISCUSSION THREAD

**Discussion by mobhuiyan98** [visgl/deck.gl#8650] [Q&A]
I'm able to produce an ellipse by extending the Scatter-plot Layer and injecting this shader code.
 
```
'vs:#decl': `
attribute float instanceMajorAxis;
attribute float instanceMinorAxis;
varying float rx;
varying float ry;
`,
        'vs:#main-end': `
rx = instanceMajorAxis * 0.01;
ry = instanceMinorAxis * 0.01;

`,
        'vs:DECKGL_FILTER_COLOR': `
float d = length(geometry.worldPosition.xy);

`,
        'fs:#decl': `
varying float rx;
varying float ry;
`,
        'fs:DECKGL_FILTER_COLOR': `
vec2 uv = abs(geometry.uv);
if ((uv.x*uv.x)/(rx*rx) + (uv.y*uv.y)/(ry*ry) > 1.0) discard;
`
```

![image](https://github.com/visgl/deck.gl/assets/66567017/50910755-f061-40e0-b46d-c87b1b5c4144)

However I'm having a really tough time figuring out how to be able to rotating the ellipse when given an angle. I'm new to WebGL , so any help on how I could implement this would be appreciated.

**Comment by Pessimistress** [comment]
You'll need to define the two axes as `vec2`s, then:

```js
'fs:#decl': `
  varying vec2 r1;
  varying vec2 r2;
`,
'fs: DECKGL_FILTER_COLOR': `
  vec2 uv = geometry.uv;
  float x = dot(uv, r1) / dot(r1, r1);
  float y = dot(uv, r2) / dot(r2, r2);
  if (x*x + y*y > 1.0) discard;
`
```

https://codepen.io/Pessimistress/pen/WNWGQyO

**Reply by TextTL** [reply]
@Pessimistress  Can you show how to  apply sizeUnits: "meters" ? ScatterPlotLayer default to meters, but the subclass seems to be "pixels"? 

Thanks,

**Reply by Pessimistress** [reply]
I don't understand your question. Are you talking about my codepen? The new attribute value is relative to the radius, so the unit is whatever radiusUnits is set to.

**Reply by TextTL** [reply]
Yes, I was referring to codepen. I tried to set sizeUnits: "meters". Zooming in/out behaves as sizeUnits: "pixels" in ScatterPlotLayer.

```
    new EllipseLayer({
      data: new Array(100),
      getFillColor: [255, 140, 0],
      getPosition: _ => [
        Math.random() - 0.5,
        Math.random() - 0.5
      ],
      getShortAxis: _ => {
        const angle = Math.PI * 2 * Math.random();
        const length = Math.random();
        return [length * Math.cos(angle), length * Math.sin(angle)];
      },
      radiusMinPixels: 20,
      opacity: 0.8,
      sizeUnits: "meters"
    })    
```

**Reply by Pessimistress** [reply]
The radius is limited by `radiusMinPixels: 20`.

**Reply by mobhuiyan98** [reply]
Would it be possible to keep the Anti-aliasing and the Outer line?
